Regex dependency inference module mapping

ryaminal opened the pull request:
PR for issue https://github.com/pantsbuild/pants/issues/19676

ryaminal commented on src/python/pants/backend/python/dependency_inference/module_mapper.py:
```diff
@@ -308,6 +311,24 @@ def providers_for_module(
+                match_pattern.sub(replace_pattern, proj_name)
+                for replace_pattern in replace_patterns
+            ]
+    return pattern_values + [fallback]
```
**question:** should the fallback be appended?
i initially started appending the fallback, which is almost always just the module with `_` e.g. `google-cloud-datastream` -> `google_cloud_datastream`. but, that is likely useless if we find a match. so perhaps this should be `return pattern_values or [fallback]` so we only use the `fallback` if no other pattern matched worked.

ryaminal replied:
done. combined with your other recommendation below.

ryaminal commented on src/python/pants/backend/python/dependency_inference/module_mapper_test.py:
```diff
@@ -59,6 +64,18 @@ def test_default_module_mapping_uses_tuples() -> None:
+            assert (
+                isinstance(replacement, FunctionType)
+                or isinstance(replacement, str)
+                or callable(replacement)
```
**comment:** this is mostly to catch partial functions
Couldn't find a concise way to capture partials(which are being used), so `callable()` should work.

ryaminal commented on src/python/pants/backend/python/dependency_inference/module_mapper_test.py:
```diff
@@ -393,14 +410,20 @@ def req(
             req("multiple_owners_types", "types-multiple_owners==1", resolve="another"),
             # Only assume it's a type stubs dep if we are certain it's not an implementation.
             req("looks_like_stubs", "looks-like-stubs-types", modules=["looks_like_stubs"]),
+            req("google-cloud-hardyhar", "google-cloud-hardyhar"),
```
**comment:** added a sample of mappings from the original list
we could add all of the mappings from the original list to feel better about the changes, but that would be a large list. instead, i added the minimum representation of the different edge cases i thought of.

ryaminal commented on src/python/pants/backend/python/dependency_inference/default_module_mapping.py:
```diff
@@ -121,80 +101,14 @@
+    # exception: kafka-python -> kafka instead of kafka_python
+    "opentelemetry-instrumentation-kafka-python": ("opentelemetry.instrumentation.kafka",),
+    # exporters
     "opentelemetry-exporter-otlp": ("opentelemetry.exporter",),
```
**comment:** opted to not extract out the `otlp` ones

ryaminal commented on src/python/pants/backend/python/dependency_inference/default_module_mapping.py:
```diff
@@ -121,80 +101,14 @@
+    # exception: kafka-python -> kafka instead of kafka_python
+    "opentelemetry-instrumentation-kafka-python": ("opentelemetry.instrumentation.kafka",),
+    # exporters
     "opentelemetry-exporter-otlp": ("opentelemetry.exporter",),
     "opentelemetry-exporter-otlp-proto-grpc": ("opentelemetry.exporter.otlp.proto.grpc",),
     "opentelemetry-exporter-otlp-proto-http": ("opentelemetry.exporter.otlp.proto.http",),
```
**comment:** opted to not extract out the `otlp` ones

ryaminal commented on src/python/pants/backend/python/dependency_inference/default_module_mapping.py:
```diff
@@ -213,25 +127,13 @@
     "pymongo": ("bson", "gridfs", "pymongo"),
     "pymupdf": ("fitz",),
     "pytest-runner": ("ptr",),
+    "python-json-logger": ("pythonjsonlogger",),
+    "python-levenshtein": ("Levenshtein",),
+    "python-lsp-jsonrpc": ("pylsp_jsonrpc",),
```
**comment:** added these back in because they don't follow the pattern, but they moved positions.
opted to not extract these out to a pattern.

ryaminal commented on src/python/pants/backend/python/dependency_inference/module_mapper_test.py:
```diff
@@ -393,14 +410,20 @@ def req(
             req("multiple_owners_types", "types-multiple_owners==1", resolve="another"),
             # Only assume it's a type stubs dep if we are certain it's not an implementation.
             req("looks_like_stubs", "looks-like-stubs-types", modules=["looks_like_stubs"]),
+            req("google-cloud-hardyhar", "google-cloud-hardyhar"),
+            req("google-cloud-secret-manager", "google-cloud-secret-manager"),
+            req("azure-keyvault-secrets", "azure-keyvault-secrets"),
+            req("django-model-utils", "model_utils"),
+            req("django-taggit", "taggit"),
+            req("opentelemetry-instrumentation-botocore", "opentelemetry-instrumentation-botocore"),
```
**comment:** added a sample of mappings from the original list
we could add all of the mappings from the original list to feel better about the changes, but that would be a large list. instead, i added the minimum representation of the different edge cases i thought of.

ryaminal commented on src/python/pants/backend/python/dependency_inference/module_mapper_test.py:
```diff
@@ -418,11 +441,85 @@ def req(
                 ),
                 "default": FrozenDict(
```
**comment:** the ordering of these are super annoying!
in general, this was a pain to test as the output doesn't easily indicate what is wrong. so, had to copy/paste the output, then format manually, then compare.
most of the time it was ordering problems. it's in alphabetical order. was tempted to wrap this in `ordered` or something but didn't.

just thought i'd share some feedback and solicit any advice/ideas on comparing nested objects so the output is better.

ryaminal commented on src/python/pants/backend/python/dependency_inference/module_mapper_test.py:
```diff
@@ -779,3 +901,40 @@ def test_issue_15111(rule_runner: RuleRunner) -> None:
+
+
+def test_generate_mappings_from_pattern_matches(rule_runner: RuleRunner) -> None:
+    assert generate_mappings_from_pattern("google-cloud-hardyhar", "bar") == [
```
**comment:** "bar"` is the fallback and will always be output, at a minimum. see a previous [comment ](https://github.com/pantsbuild/pants/pull/19684/files#r1309078101) on thoughts about including the fallback.

ryaminal commented on src/python/pants/backend/python/dependency_inference/default_module_mapping.py:
```diff
@@ -3,87 +3,81 @@
+from typing import Match
+
+
+def all_hyphen_to_dot(m: Match) -> str:
```
**comment:** functions for reusability?
tried to be very verbose with the names so these can easily be used in the future for other pattern matches.
opted to not use `lambdas`, because they are anonymous, or simple "string" repls because i thought they'd be less likely to be reused in the future.

ryaminal commented on src/python/pants/backend/python/dependency_inference/default_module_mapping.py:
```diff
@@ -3,87 +3,81 @@
+google-cloud-foo will be used.
+
+The match is either a string or a function that takes a re.Match and returns
+the replacement. see re.sub for more information
```
**comment:** [docs](https://docs.python.org/3/library/re.html#re.sub) for re.sub

ryaminal commented on src/python/pants/backend/python/dependency_inference/module_mapper.py:
```diff
@@ -308,6 +311,24 @@ def providers_for_module(
+    """
+    pattern_values = []
+    for match_pattern, replace_patterns in DEFAULT_MODULE_PATTERN_MAPPING.items():
+        if match_pattern.match(proj_name) is not None:
```
**question:** stop after the first match.
right now if a particular project matches multiple patterns, then all of them will be applied.
curious what folks thoughts are on this?
an example:
```python
{
  re.compile("bob"): [do_something],
  re.compile(".*bob.*"): [do_another_something],
}
```
in this case we'd get replacement patterns that are generate by both `do_something` and `do_another_something`. curious what folks thoughts are on this behavior.

if the patterns are identical like:
```python
{
  re.compile("bob"): [do_something],
  re.compile("bob"): [do_another_something],
}
```
then only one will be done and i believe it's the last one. this is because `re.compile("bob")` always generates to the same key so the dict will only have one in there.

huonw replied:
Hm, good question. In some sense, it seems like it should be an error in the module mappings if patterns overlapped, so I'd be fine to just stop after the first match. 

I don't know if it's worth the effort to actually validate non-overlapping-ness in a pants test. I imagine one could theoretically intersect the regexes and then see if any match exists, e.g. with https://pypi.org/project/interegular/ .

huonw commented on src/python/pants/backend/python/dependency_inference/module_mapper.py:
```diff
@@ -361,7 +382,12 @@ def add_modules(modules: Iterable[str], *, type_stub: bool = False) -> None:
                 add_modules(stub_modules, type_stub=True)
             else:
-                add_modules(DEFAULT_MODULE_MAPPING.get(proj_name, (fallback_value,)))
+                add_modules(
+                    DEFAULT_MODULE_MAPPING.get(
+                        proj_name,
+                        generate_mappings_from_pattern(proj_name, fallback_value),
+                    )
+                )
```
Related to the comment above, it might make sense to move this into a separate function like:

```python
def generate_mapping(prod_name, fallback_value) -> Iterable[str]:
    if from_mapping := DEFAULT_MODULE_MAPPING.get(proj_name):
        return from_mapping

    if from_patterns := generate_mappings_from_pattern(proj_name):
        return from_patterns
    
    return (fallback_value,)
```

And then this becomes

```suggestion
                add_modules(generate_mappings(proj_name, fallback_value))
```

ryaminal replied:
did a version of this. good suggestion.

huonw commented on src/python/pants/backend/python/dependency_inference/module_mapper_test.py:
```diff
@@ -779,3 +903,48 @@ def test_issue_15111(rule_runner: RuleRunner) -> None:
     )
+
+
+def test_generate_mappings_from_pattern_matches(rule_runner: RuleRunner) -> None:
```
What do you think about writing this as a parametrised test:

```python
@pytest.mark.parametrize(
    ("proj_name", "expected_modules")
    [
        ("google-cloud-hardyhar", ["google.cloud.hardyhar", ...]),
        ...
    ]
)
def test_generate_mappings_from_pattern_matches(proj_name: str, expected_modules: list[str]) -> None:
    assert generate_mappings_from_pattern(proj_name, "bar") == expected_modules
```

I find is a slightly nicer experience because one can see all the failures at once.

(Also, I'd expect the `rule_runner` fixture is only required in tests that actually call methods on it, e.g. it's might not be needed in this one and `test_generate_mappings_from_pattern_no_match`. Have you found otherwise?)

ryaminal replied:
ah, parameterized is much better. will do that.

and i think you are correct that the `rule_runner` is not needed. likely a copy-paste thing i did and didn't bother to fix.

huonw reviewed (commented):
Thanks for working on this @ryaminal. I've reviewed this at a functional level, but not at a "directional" level, so maybe wait for @benjyw's input before devoting too much effort to my review.

benjyw reviewed (commented):
This is a great idea. I am just wondering about the tradeoff between complexity and benefit. This file was once dirt-simple, and is now somewhat confusing. Is there an 80/20 thing here where we could greatly simplify the implementation at the cost of needing a few more manual cases. E.g., what if we didn't handle the `_v1` suffix etc. at all, and we still needed to enumerate those cases as before?

It seems like at least the azure, django and opentelemetry cases are really simple, for example? And most, but not all, of the google ones?

ryaminal commented on src/python/pants/backend/python/dependency_inference/default_module_mapping.py:
```diff
@@ -3,87 +3,84 @@
+"""
+DEFAULT_MODULE_PATTERN_MAPPING: Dict[re.Pattern, Iterable] = {
+    re.compile(r"""^(google-cloud-)([^.]+)"""): [
+        two_groups_hyphens_two_replacements_with_suffix,
+        partial(two_groups_hyphens_two_replacements_with_suffix, custom_suffix="_v1"),
+        partial(two_groups_hyphens_two_replacements_with_suffix, custom_suffix="_v2"),
+        partial(two_groups_hyphens_two_replacements_with_suffix, custom_suffix="_v3"),
```
that is correct, yeah. decided to not to beta/alpha ones and only "released" versions. the highest i saw was `v3`. only like 2 packages at v3. 90% are at v1, i think. haven't run the numbers but yeah, v1 is the most common.

speaking of this, does pants/pex handle `from google.cloud.aiplatform_v3 import AIPlatform` sort of patterns? i assumed "no" and changed my local code to all be `import google.cloud.aiplatform_v3` stuff.

huonw replied:
Pants handles that one, and also `from google.cloud import aiplatform_v3` too.

ryaminal commented:
@benjyw, you are certainly correct that doing it this way(with regex and functions and tests and potentially breaking) adds complexity and potentially confusion.

i'm hoping that the complexity here is outweighed by the frustration of having to either add mappings to my pants project or to the default mapping. this was mostly spurred on that i was annoyed that i have to add this somewhere when the pattern is "guessable". i also have some convincing to do on my team and at my org that pants is the right choice, so easing that adoption by removing one more hurdle seemed promising.

on to your question about an 80/20 compromise that keeps the default mapping simple but still covers more use cases. the `google.cloud` and `opentelemetry` are the most complex with `google.cloud` being more complex. i don't know, however, how we can easily handle `google.cloud` without someone either adding tons of stuff to `default_mapping` or lots of people adding a few things to their local `mapping`.

we could potentially simplify it by "forcing" folks to follow the `from google.cloud import <stuff>` pattern. might simplify some things but... hmmm.

could also just discard this and do the lock-based [solution](https://github.com/pantsbuild/pants/issues/19676#issuecomment-1699563971) that @kaos mentioned.

ryaminal commented:
hey all. before i go through and make the changes that @huonw has recommended, want to make sure there is some alignment on going forward.
even if the conclusion is, "nah, hardpass on this change".

benjyw commented:
Hi, I like the change overall, I'm just concerned about its complexity. Maybe all this needs to assuage my feelings is more docstrings. E.g., the docstring for two_groups_hyphens_two_replacements_with_suffix is good (explains what the function does, and gives an example). Maybe the other functions could have equally descriptive docstrings?

ryaminal commented:
@benjyw, makes sense. i made the recommended changes and simplified a few things. added some docstrings that hopefully explain things a little better.

ryaminal commented:
hmmm, lint is failing in ci/cd and local if i run it on everything `pants lint check ::`. not certain if this is because of something changed in here, but it very well could be. looking in to it but if anyone has any thoughts...

benjyw reviewed (approved):
Thanks! With the new, detailed docstrings I am satisfied re the complexity, and the test speaks to correctness.

benjyw commented:
@huonw any comments before I merge?

huonw reviewed (approved):
I like it. Thanks for iterating!

huonw commented:
Thanks @ryaminal! `main` is currently prepping for 2.19, so this will be available in that release.

ryaminal commented:
awesome. thanks for all of the help with this!
